libweb3jsonrpc/Test.cpp

```diff
@@ -33,6 +34,64 @@ using namespace jsonrpc;
 
 Test::Test(eth::Client& _eth): m_eth(_eth) {}
 
+Json::Value fillJsonWithState(eth::State const& _state)
+{
+    Json::Value oState;
+    for (auto const& a : _state.addresses())
+    {
+        Json::Value o;
+        o["balance"] = toCompactHexPrefixed(_state.balance(a.first), 1);
+        o["nonce"] = toCompactHexPrefixed(_state.getNonce(a.first), 1);
+
+        Json::Value store;
+        for (auto const& s : _state.storage(a.first))
+            store[toCompactHexPrefixed(s.second.first, 1)] =
+                toCompactHexPrefixed(s.second.second, 1);
+        o["storage"] = store;
+
+        if (_state.code(a.first).size() > 0)
+            o["code"] = toHexPrefixed(_state.code(a.first));
+        else
+            o["code"] = "";
+
+        oState[toHexPrefixed(a.first)] = o;
+    }
+    return oState;
+}
+
+string exportLog(eth::LogEntries const& _logs)
+{
+    RLPStream s;
+    s.appendList(_logs.size());
+    for (eth::LogEntry const& l : _logs)
+        l.streamRLP(s);
+    return toHexPrefixed(sha3(s.out()));
+}
+
+const int c_postStateJustHashVersion = 1;
+const int c_postStateFullStateVersion = 2;
+const int c_postStateLogHashVersion = 3;
+string Test::test_getPostState(Json::Value const& param1)
+{
+    Json::Value out;
+    Json::FastWriter fastWriter;
+    if (u256(param1["version"].asString()) == c_postStateJustHashVersion)
+        return toJS(m_eth.postState().state().rootHash());
+    else if (u256(param1["version"].asString()) == c_postStateFullStateVersion)
+    {
+        out = fillJsonWithState(m_eth.postState().state());
+        return fastWriter.write(out);
+    }
+    else if (u256(param1["version"].asString()) == c_postStateLogHashVersion)
+    {
+        if (m_eth.blockChain().receipts().receipts.size() != 0)
+            return exportLog(m_eth.blockChain().receipts().receipts[0].log());
+        else
+            return toJS(EmptyListSHA3);
+    }
+    return "";
```

Comment:
how do we report an error via RPC reply ?

Reply:
I don't know, look at other methods.

Reply:
well there is ```BOOST_THROW_EXCEPTION(JsonRpcException(Errors::ERROR_RPC_INVALID_PARAMS));```

but it stops the client. which is not a good thing. you dont want a client to stop on malicious rpc

Reply:
It shouldn't stop the client, all RPC methods report errors kind of like that and the last time I tried it I saw correct error message on the caller side.

Could you investigate whether it crashes because of unhandled exception or for other reason?

Reply:
Also related issue is https://github.com/ethereum/cpp-ethereum/issues/4726

Reply:
no. it does not crash the client. I was wrong. the rpc error is returned.